src/dex/wombat/wombat-bmw.ts

```diff
@@ -107,37 +110,68 @@ export class WombatBmw extends StatefulEventSubscriber<BmwState> {
         .call({}, blockNumber)
     ).returnData;
 
-    const lpTokens = returnData.map(
-      (data: BytesLike) =>
-        WombatBmw.bmwInterface.decodeFunctionResult('poolInfo', data)[0],
+    const lpTokens = returnData.map((data: BytesLike) =>
+      WombatBmw.bmwInterface
+        .decodeFunctionResult('poolInfo', data)[0]
+        .toLowerCase(),
     );
 
-    // 2. get all pools and underlying tokens
-    inputs = lpTokens.map((lpToken: string) => ({
-      target: lpToken,
-      callData: WombatBmw.assetInterface.encodeFunctionData('pool'),
-    }));
+    inputs = [];
+    for (const lpToken of lpTokens) {
+      inputs.push({
+        target: lpToken,
+        callData:
+          WombatBmw.assetInterface.encodeFunctionData('underlyingToken'),
+      });
+      inputs.push({
+        target: lpToken,
+        callData: WombatBmw.assetInterface.encodeFunctionData('pool'),
+      });
+    }
 
     returnData = (
       await this.dexHelper.multiContract.methods
         .aggregate(inputs)
         .call({}, blockNumber)
     ).returnData;
 
-    const pools = returnData
-      .map(
-        (data: BytesLike) =>
-          WombatBmw.assetInterface.decodeFunctionResult('pool', data)[0],
-      )
-      .map((pool: string) => pool.toLowerCase());
+    const pool2AssetInfo = new Map<Address, Map<Address, Address>>();
+    let i = 0;
+    for (const lpToken of lpTokens) {
+      const underlyingToken = WombatBmw.assetInterface
+        .decodeFunctionResult('underlyingToken', returnData[i++])[0]
+        .toLowerCase();
+      const pool = WombatBmw.assetInterface
+        .decodeFunctionResult('pool', returnData[i++])[0]
+        .toLowerCase();
+
+      if (!pool2AssetInfo.has(pool)) {
+        pool2AssetInfo.set(pool, new Map<Address, Address>());
+      }
+      pool2AssetInfo.get(pool)!.set(lpToken, underlyingToken);
+    }
 
-    const promises = [];
-    for (const pool of pools) {
-      if (!bmwState.pools.includes(pool)) {
+    inputs = [];
+    pool2AssetInfo.forEach((asset2TokenMap, pool) => {
+      inputs.push({
+        target: pool,
+        callData: WombatBmw.poolInterface.encodeFunctionData(
+          'creditForTokensHaircut',
+        ),
+      });
+    });
+
+    returnData = await this.dexHelper.multiContract.methods
+      .tryAggregate(false, inputs)
+      .call({}, blockNumber);
+    i = 0;
+    pool2AssetInfo.forEach((asset2TokenMap, pool) => {
+      // don't add cross-chain pools
+      if (!returnData[i++].success) {
```

Comment:
Q: why not add cross chain pool? Some chain like avax only has cross chain pools.

Reply:
Fixed. Cross-chain pool is added back now.